File: cases/taint/scala/reflective-invocation-positive/ReflectiveInvocationPositive.scala
package dataflowbench

object ReflectiveInvocationPositive {
  class Target {
    def leak(value: String): Unit = {
      dfb_sink(value)
    }

    def drop(value: String): Unit = {
      dfb_sink("clean")
    }
  }

  def dfb_source(): String = { // DFB-SOURCE: reflective-invocation-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: reflective-invocation-sink

  def run(): Unit = {
    val target = new Target()
    val name = "leak"
    val method = classOf[Target].getMethod(name, classOf[String]) // DFB-WITNESS: reflective-invocation-resolve
    method.invoke(target, dfb_source())
  }
}

File: cases/taint/php/reflective-invocation-positive/reflective_invocation_positive.php
<?php
class Target
{
    public function leak(string $value): void
    {
        dfb_sink($value);
    }

    public function drop(string $value): void
    {
        dfb_sink("clean");
    }
}

function dfb_source(): string { // DFB-SOURCE: reflective-invocation-input
    return "tainted";
}

function dfb_sink(string $value): void {} // DFB-SINK: reflective-invocation-sink

function run(): void {
    $target = new Target();
    $name = "leak";
    $target->$name(dfb_source()); // DFB-WITNESS: reflective-invocation-resolve
}
